niceman/resource/session.py

```diff
@@ -375,19 +521,57 @@ def read(self, path):
     def mkdir(self, path, parents=False):
         """Create a directory
         """
-        self.execute_command(["mkdir"] + (["-p"] if parents else [""]) + [path])
+        command = ["mkdir"]
+        if parents: command.append("-p")
+        command += [path]
+        self.execute_command(command)
 
         if not self.isdir(path):
             raise CommandError(cmd='mkdir', msg="Failed to create directory")
 
     def isdir(self, path):
         try:
-            self.execute_command(['test', '-d', path])
+            out, err = self.execute_command(['test', '-d', path])
+        except Exception as exc:  # TODO: More specific exception?
+            lgr.debug("Check for directory failed: %s", exc_str(exc))
+            return False
+        if not err:
             return True
-        except CommandError:
+        else:
+            lgr.debug("Standard error was not empty (%r), thus assuming that "
+                      "test for directory has failed", err)
             return False
-    # chmod?
-    # chown?
+
+    def chmod(self, path, mode, recursive=False):
+        """Set the mode of a remote path
+        """
+        command = ['chmod']
+        if recursive: command += ["-R"]
+        command += [mode] + [path]
+        self.execute_command(command)
+
+    def chown(self, path, uid=-1, gid=-1, recursive=False):
+        """Set the user and gid of a path
+        """
+        command = ['chown']
+        if recursive: command += ["-R"]
+        if int(uid) > 0 and int(gid) > 0: command += ["{}.{}".format(uid, gid)]
+        elif int(uid) > -1: command += [uid]
+        elif int(uid) > -1: command = ['chgrp'] + [gid]
+        else: raise CommandError(cmd='chown', msg="Invalid command \
+            parameters.")
+        command += [path]
+        self.execute_command(command)
+
+    def local_chown(self, src_path, dest_path, uid=-1, gid=-1):
+        target = dest_path + os.sep + os.path.basename(src_path)
+        os.chown(target, int(uid), int(gid))
+        if os.path.isdir(target):
+            for root, dirs, files in os.walk(target):
+                for dir in dirs:
+                    os.chown(os.path.join(root, dir), int(uid), int(gid))
+                for file in files:
+                    os.chown(os.path.join(root, file), int(uid), int(gid))
```

Comment:
`file` and `dir` are internal keywords of python.. ideally do not overload. disambiguate with `_`, e.g. `file_` and `dir_`.  I know that it is minor -- but believe me, I did have to fix bugs where I had to stare at the screen for a while to figure out that `file` is not actually a Python `file`